.github/github_workflow_scripts/purge_branch_protection_rules.py

```diff
@@ -0,0 +1,358 @@
+
+
+# Queries
+GET_BRANCH_PROTECTION_GRAPHQL_QUERY_TEMPLATE = """query($owner: String!, $name: String!) {
+repository(owner: $owner, name: $name) {
+    branchProtectionRules(first: 100) {
+    nodes {
+        id
+        pattern
+        matchingRefs(first: 10) {
+        totalCount
+        }
+    }
+    }
+}
+}"""
+
+DELETE_BRANCH_PROTECTION_RULE_QUERY_TEMPLATE = """mutation deleteBranchProtectionRule($branchProtectionRuleId: ID!) {
+    deleteBranchProtectionRule(input: {branchProtectionRuleId: $branchProtectionRuleId}) {
+        clientMutationId
+    }
+}"""
+
+
+# Helper Functions
+
+def get_repo_owner_and_name() -> tuple[str, str]:
+    """
+    Extracts the repository owner and name from a given repository string.
+
+    Returns:
+        A `Tuple[str, str]` containing the repository owner and name.
+
+    Raises:
+        `ValueError`: If the input string is not in the expected 'owner/repository' format.
+    """
+
+    repo = os.getenv(GH_REPO_ENV_VAR, DEFAULT_REPO)
+
+    parts = repo.split('/')
+
+    if len(parts) != 2:
+        raise ValueError("Input string must be in the format 'owner/repository'.")
+
+    owner, name = parts
+    return owner, name
+
+
+def convert_response_to_bpr(response: dict[str, Any]) -> list[BranchProtectionRule]:
+    """
+    Helper method to convert the response to a list
+    of `BranchProtectionRule`.
+
+    Arguments:
+    - `response` (``dict[str, Any]``): The response data.
+
+    Returns:
+    - a `list[BranchProtectionRule]`.
+
+    Raises:
+    - `KeyError | AttributeError` in case the conversion fails.
+    """
+
+    rules: list[BranchProtectionRule] = []
+
+    try:
+        for node in response.get('data', {}).get('repository').get('branchProtectionRules').get('nodes'):
+            rule = BranchProtectionRule(
+                id=node.get("id"),
+                pattern=node.get("pattern"),  # ignore: type
+                matching_refs=node.get("matchingRefs").get("totalCount")
+            )
+
+            rules.append(rule)
+    except (KeyError, AttributeError) as e:
+        raise e.__class__(f"{e.__class__.__name__} parsing '{response=}' as a branch protection rule: {e}")
+
+    return rules
+
+
+def should_delete_rule(rule: BranchProtectionRule) -> bool:
+    """
+    Check whether we should delete this rule.
+    To determine if we should delete the rule we check that:
+
+    * The rule is not in the list of protected rules (see `self.PROTECTED_RULES`)
+    * The rule does not apply to any branches.
+
+    Returns:
+    - `True` if we can delete the rule, `False` otherwise.
+    """
+
+    should = False
+
+    if rule.pattern in PROTECTED_RULES:
+        logger.info(f"{rule} not deleted because it's in the list of protected rules '{','.join(PROTECTED_RULES)}'")
+    elif rule.matching_refs > 0:
+        logger.info(f"{rule} not deleted because it's associated to {rule.matching_refs} existing branches/refs")
+    else:
+        should = True
+
+    return should
```

Comment:
```suggestion
    if rule.pattern in PROTECTED_RULES:
        logger.info(f"{rule} not deleted because it's in the list of protected rules '{','.join(PROTECTED_RULES)}'")
        return False
    if rule.matching_refs > 0:
        logger.info(f"{rule} not deleted because it's associated to {rule.matching_refs} existing branches/refs")
        return False
    logging.info(f"Rule is deleted...")
    return True
```
multiple return statements is not a bad practice.